ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/all_reduce_pybind.cpp

```diff
@@ -0,0 +1,92 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "all_reduce_pybind.hpp"
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "ttnn/cpp/pybind11/decorators.hpp"
+#include "ttnn/operations/ccl/ccl_host_datastructures.hpp"
+#include "ttnn/operations/experimental/ccl/all_reduce/all_reduce.hpp"
+#include "ttnn/types.hpp"
+
+#include "ttnn/operations/reduction/generic/generic_reductions.hpp"
+
+namespace ttnn::operations::experimental::ccl {
+
+namespace detail {
+
+template <typename ccl_operation_t>
+void bind_all_reduce(pybind11::module& module, const ccl_operation_t& operation, const char* doc) {
+    bind_registered_operation(
+        module,
+        operation,
+        doc,
+        ttnn::pybind_overload_t{
+            [](const ccl_operation_t& self,
+               const ttnn::Tensor& input_tensor,
+               ttnn::operations::reduction::ReduceType math_op,
+               const uint32_t num_links,
+               const ttnn::MemoryConfig& memory_config,
+               ttnn::ccl::Topology topology,
+               const std::optional<size_t> num_workers,
+               const std::optional<size_t> num_buffers_per_channel) -> ttnn::Tensor {
+                return self(input_tensor, math_op, num_links, memory_config, topology, num_workers, num_buffers_per_channel);
+            },
+            py::arg("input_tensor"),
+            py::arg("math_op"),
+            py::kw_only(),
+            py::arg("num_links") = 1,
+            py::arg("memory_config") = std::nullopt,
+            py::arg("topology") = ttnn::ccl::Topology::Ring,
+            py::arg("num_workers") = std::nullopt,
+            py::arg("num_buffers_per_channel") = std::nullopt});
+}
+
+}  // namespace detail
+
+
+void py_bind_all_reduce(pybind11::module& module) {
+
+    detail::bind_all_reduce(
+        module,
+        ttnn::experimental::all_reduce,
+        R"doc(
+
+        Performs an all_reduce operation on multi-device :attr:`input_tensor` across all devices.
+
+        Args:
+            input_tensor (ttnn.Tensor): multi-device tensor
+            dim (int): Dimension to perform operation
+
+        Keyword Args:
+            num_links (int, optional): Number of links to use for the all-gather operation. Defaults to `1`.
+            memory_config (ttnn.MemoryConfig, optional): Memory configuration for the operation. Defaults to `input tensor memory config`.
```

Comment:
Random thought (not specific feedback to this PR), but maybe we can have "standard" descriptions for some of these standard args and just include them as a part of the formatted string...? This should really help us have uniform descriptions!